src/validate-icon.c

```diff
@@ -27,46 +28,82 @@
 #include <gdk-pixbuf/gdk-pixbuf.h>
 #include <glib/gstdio.h>
 #include <errno.h>
+#include <fcntl.h>
 #include <unistd.h>
 
+#include "xdp-utils.h"
+
 #ifdef __FreeBSD__
 #define execvpe exect
 #endif
 
 #define ICON_VALIDATOR_GROUP "Icon Validator"
 #define MAX_ICON_SIZE 512
 #define MAX_SVG_ICON_SIZE 4096
+#define BUFFER_SIZE 4096
 
 static int
-validate_icon (const char *filename)
+validate_icon (int input_fd)
 {
-  GdkPixbufFormat *format;
-  int max_size;
-  int width, height;
-  g_autofree char *name = NULL;
   const char *allowed_formats[] = { "png", "jpeg", "svg", NULL };
-  g_autoptr(GdkPixbuf) pixbuf = NULL;
+  g_autoptr(GBytes) bytes = NULL;
   g_autoptr(GError) error = NULL;
+  GdkPixbufFormat *format;
   g_autoptr(GKeyFile) key_file = NULL;
   g_autofree char *key_file_data = NULL;
+  g_autoptr(GdkPixbufLoader) loader = NULL;
+  g_autoptr(GMappedFile) mapped = NULL;
+  int max_size, width, height;
+  g_autofree char *name = NULL;
+  GdkPixbuf *pixbuf;
 
-  format = gdk_pixbuf_get_file_info (filename, &width, &height);
-  if (format == NULL)
+  /* Ensure that we read from the beginning of the file */
+  lseek (input_fd, 0, SEEK_SET);
+
+  if (!(mapped = g_mapped_file_new_from_fd (input_fd, FALSE, &error)))
     {
-      g_printerr ("Format not recognized\n");
+      g_printerr ("Failed to create mapped file for image: %s\n", error->message);
       return 1;
     }
 
-  if (width != height)
+  bytes = g_mapped_file_get_bytes (mapped);
+
+  if (g_bytes_get_size (bytes) == 0)
     {
-      g_printerr ("Expected a square icon but got: %dx%d\n", width, height);
+      g_printerr ("Image is 0 bytes\n");
+      return 1;
+    }
+
+  loader = gdk_pixbuf_loader_new ();
+
+  if (!gdk_pixbuf_loader_write_bytes (loader, bytes, &error) ||
+      !gdk_pixbuf_loader_close (loader, &error) ||
+      !(pixbuf = gdk_pixbuf_loader_get_pixbuf (loader)))
+    {
+      g_printerr ("Failed to load image: %s\n", error->message);
+      gdk_pixbuf_loader_close (loader, NULL);
+      return 1;
+    }
```

Comment:
Hm, I think this will eventually call `gdk_pixbuf_loader_close (loader, NULL);` too much. I think these checks have to be split after all.

```suggestion
  if (!gdk_pixbuf_loader_write_bytes (loader, bytes, &error) ||
      !gdk_pixbuf_loader_close (loader, &error))
    {
      g_printerr ("Failed to load image: %s\n", error->message);
      gdk_pixbuf_loader_close (loader, NULL);
      return 1;
    }
    
  pixbuf = gdk_pixbuf_loader_get_pixbuf (loader);
  if (!pixbuf)
    {
      g_printerr ("Failed to load image: %s\n", error->message);
      return 1;
    }
```

Reply:
`_close()` checks whether the loader was already closed. https://gitlab.gnome.org/GNOME/gdk-pixbuf/-/blob/master/gdk-pixbuf/gdk-pixbuf-loader.c#L839

I still changed it since it makes it easier to read the code.